<a href="https://colab.research.google.com/github/unt-iialab/INFO5731_Spring2020/blob/master/Assignments/INFO5731_Assignment_One.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **INFO5731 Assignment One**

This assignment includes range of programming tasks and data analysis exercises, aiming to enhance your Python skills and proficiency in string manipulation, numerical computations, and data science. Follow the instructions carefully and design well-structured Python programs to address each question.

Late Submission will have a penalty of 10% reduction for each day after the deadline.

# Question 1 (10 points).

The formula that describes the volume V of a hemisphere with radius r is the following:

$ V=\frac{2}{3}\ast\ \pi\ast\ r^3 $

Write a program to calculate the value of V when r is in the range of even numbers from 1-20. Output the result in the following format:

	r	V
	2	…
	4	…
	6	…
	.
	.
	.
	20	…


In [1]:
# Your code here
print("r\tV")
pi=3.14159265359
# Setting upper limit in the range() as 21 because in for loop the upper limit will not be included.
for i in range(2,21,2):
  print(i,2/3*pi*i**3)



r	V
2 16.755160819146667
4 134.04128655317334
6 452.38934211696
8 1072.3302924253867
10 2094.395102393333
12 3619.11473693568
14 5747.020160967307
16 8578.642339403094
18 12214.51223715792
20 16755.160819146666


# Question 2 (20 points)

(a) Write a Python program to generate a sequence of 12 random numbers falling within the range of 50 to 150.

(b) Arrange the generated numbers in ascending order. Display the sorted list.

(c) Extend the program to arrange the numbers in descending order as well. Present the sorted list accordingly.

(d) Compute and output the average of the entire group of generated numbers.

(e) Enhance the program to calculate and display the average of the first 6 numbers from both the ascending and descending sorted lists.

In [2]:
# Write code for each of the sub parts with proper comments.
import random
# (a)
# getting random values in range 50-150
random_list = random.sample(range(50, 150), 12)
print("12 random numbers between 50,150 ",random_list)
# (b)
# sorting (assending order)
random_list.sort()
print("Sorted list in assending order ",random_list)
# (c)
# sorting descending order
random_list.sort(reverse=True)
print("Sorted list in descending order ",random_list)
# (d)
total_sum =sum(random_list)
# average
average=total_sum/len(random_list)
print("Average of the list ",average)
# (e)
random_list.sort()
average_ascending=0
average_descending=0
#calculating average of first 6 numbers in ascending ordered list
for i in range(6):
  average_ascending=average_ascending+random_list[i]
print("Average of first 6 numbers in ascending ordered list ",average_ascending/6)
random_list.sort(reverse=True)
#calculating average of first 6 numbers in descending ordered list
for i in range(6):
  average_descending=average_descending+random_list[i]
print("Average of first 6 numbers in descending ordered list ",average_descending/6)







12 random numbers between 50,150  [86, 98, 145, 105, 78, 116, 144, 103, 120, 70, 106, 77]
Sorted list in assending order  [70, 77, 78, 86, 98, 103, 105, 106, 116, 120, 144, 145]
Sorted list in descending order  [145, 144, 120, 116, 106, 105, 103, 98, 86, 78, 77, 70]
Average of the list  104.0
Average of first 6 numbers in ascending ordered list  85.33333333333333
Average of first 6 numbers in descending ordered list  122.66666666666667


# Question 3 (20 points).

Create a Python program that encompasses various string manipulation task:

(a) Prompt the user to input two sentences.

(b) Transform each sentence into a list, where each element corresponds to a word in the sentence. For example, if the sentences are "Hello, how are you doing today?" and "This is an example sentence for the program!"

The program should output:
['Hello', ',', 'how', 'are', 'you', 'doing', 'today', '?', 'This', 'is', 'an', 'example', 'sentence', 'for', 'the', 'program', '!']

(c) Compare the two sentences. Print out a list of words that occur in both sentences.

(d) Remove any punctuation from the sentence and print the sanitized sentences. It should remove at least the following characters, but it can remove more: period(“.”), comma (“,”), semicolon (“;”), and colon (“:”).

(e) Count the number of vowels in each of the sentences. Print out each vowel and the occurrences  in the each of the two sentences separately, such as:

Sentence1: a:2, e:2, i:0, etc.
Sentence2: a:1, e:2, i:2, etc.

Design a well-structured Python program to execute these tasks efficiently.

In [ ]:
from string import punctuation
# (a)
import re
 # taking inputs
input1=input("Enter first sentence: ")
input2=input("Enter second sentence: ")


# (b)
#splitting into words
def splitting_sentence(sentence):
  words=sentence.split()
  res=[]
  for word in words:
    if word[-1] in ".,?;'-#@!:":
      res.append(word[:-1])
      res.append(word[-1])
    else:
      res.append(word)
  return res
list1=splitting_sentence(input1)
list2=splitting_sentence(input2)
list3=list1+list2
print(list3)


# (c)
common_words=[]
for word in list1:
  if word.lower() in (w.lower() for w in list2):
    common_words.append(word)
print("Common words in both sectences: ",common_words)


# (d)
def remove_punctuation(sentence):
  # removing punctuations
  punctuations = ".;':,?{}()[]"
  for letter in punctuations:
    sentence = sentence.replace(letter, "")
  return sentence
print("sentence1 with out punctuation: ",remove_punctuation(input1))
print("sentence2 with out punctuation: ",remove_punctuation(input2))


# (e)
def count_vowels(sentence):
  vowels="aeiouAEIOU"
  count={}
  # counting vovels
  for letter in vowels:
    count[letter]=sentence.count(letter)
  return count
print("Count of vovels in sentence1: ",count_vowels(input1))
print("Count of vovels in sentence1: ",count_vowels(input2))








Enter first sentence: I like problem solving.
Enter second sentence: I know python.
['I', 'like', 'problem', 'solving', '.', 'I', 'know', 'python', '.']
Common words in both sectences:  ['I', '.']
sentence1 with out punctuation:  I like problem solving
sentence2 with out punctuation:  I know python
Count of vovels in sentence1:  {'a': 0, 'e': 2, 'i': 2, 'o': 2, 'u': 0, 'A': 0, 'E': 0, 'I': 1, 'O': 0, 'U': 0}
Count of vovels in sentence1:  {'a': 0, 'e': 0, 'i': 0, 'o': 2, 'u': 0, 'A': 0, 'E': 0, 'I': 1, 'O': 0, 'U': 0}


# Question 4 (20 Points)

Design and implement a simple command-line calculator program. The program should handle both integers and floating-point numbers.

(a) Write program that takes a mathematical expression as input and evaluates it. The program should support addition, subtraction, multiplication, and division.

(b) Extend the calculator to handle parentheses in mathematical expressions. Implement the necessary logic to evaluate expressions with nested parentheses.

(c) Enhance your calculator to include additional mathematical operations such as exponentiation, square root, and modulus.

(d) Implement error handling in the calculator program. Handle cases such as division by zero and invalid input expressions. Provide meaningful error messages.

(e) Implement a command to exit the calculator.

In [ ]:
import math

def evaluate_expression(expression):
  nums=expression.split()
  # setting format if len is 2, format: operator number.

  if len(nums)==2:
    op,num1=nums
    num1=float(num1)
    if op=="sqrt":
      if num1<0:
        return "Error: No square root for negatives"
      return math.sqrt(num1)
    elif op=="exp":
        return math.exp(num1)
    else:
        return "Error: Unsupported operator"
# setting format if len is 3, format: number operator number.
  if len(nums)==3:
    num1,op,num2=nums
    num2=float(num2)
    num1=float(num1)
    if op=="+":
      return num1+num2
    elif op=="-":
      return num1-num2
    elif op=="*":
      return num1*num2
    elif op=="/":
      if num2==0:
        return "Error: Cannot do division by zero"
      return num1/num2
    elif op=="%":
      if num2==0:
        return "Error: Cannot do modulo by zero"
      return num1%num2
    elif op=="**":
      return num1**num2
    elif op == "sqrt":
      if num1 < 0:
          return "Error: Cannot calculate the square root of a negative number."
      return math.sqrt(num1)
    elif op == "exp":
      return math.exp(num1)
    else:
      return "Enter a valid operator"
  else:
    return "Invalid expression, use format: operand operator operand (e.g., 2 + 3)"

def calculator(expression):
  try:
      # checking paranthesis
    while '(' in expression:
      right_most_open_bracket=expression.rfind('(')
      close_for_right_most_open_bracket=expression.find(')',right_most_open_bracket)
      if close_for_right_most_open_bracket==-1:
        return "Error: Enter a valid paranthesis"
      inner_expression=expression[right_most_open_bracket+1:close_for_right_most_open_bracket]
      inner_result=str(eval(inner_expression))
      expression=expression[:right_most_open_bracket]+inner_result+expression[close_for_right_most_open_bracket+1:]
    return evaluate_expression(expression)
  except Exception as e:
    return e
print("You can add paranthesis in this calculator and also try operators such as modulus(%),exp,sqrt")
print("Enter 'quit' to exit the calculator")
while True:
  expression=input("Enter the expression: ")
  # use 'quit' to quit the calculator
  if expression.lower()=="quit":
    break
  print(calculator(expression))



You can add paranthesis in this calculator and also try operators such as modulus(%),exp,sqrt
Enter 'quit' to exit the calculator
Enter the expression sqrt 16
4.0
Enter the expression 2-(3+1)
Invalid expression, use format: operand operator operand (e.g., 2 + 3)
Enter the expression 2 - (3 + 1)
-2.0
Enter the expression exp 2
7.38905609893065
Enter the expression 4 / 0
Error: Cannot do division by zero
Enter the expression quit


# Question 5 (30 points).

In the field of Data Science, data is often formatted as a comma-delimited (CSV) file, where each line in a file corresponds to a field's value. Refer to Chapter 9 (p. 106) for more information on comma-delimited files.

To answer this question, you will need to download the file [Assignment1_denton_housing.csv](https://raw.githubusercontent.com/unt-iialab/info5731-spring2022/main/assignments/Assignment1_denton_housing.csv) included with this assignment. The file contains statistics about housing in the city of Denton from 2008 to 2014. Write a program to implement the following questions.

(a) (5 pts) Calculate the difference in the number of Occupied Housing Units from year to year and print it. The difference must be calculated for the consecutive years such as 2008-2009, 2009-2010 etc. Finally, print the values in the ascending order.

(b) (5 pts) For all the years, calculate the percentage of housing units which are vacant an occupied. Print the results in the following format:

Year Vacant Housing Units Occupied Housing Units

    2008  30%   70%                          
	2009 ----- -----
	2010 ----- -----
    2011 ----- -----
	2012 ----- -----
    2013 ----- -----
	2014 ----- -----
  
  

(c) (10 pts) Calculate and print the valued and years in which the highest number of housing units were vacant and occupied. Print the results in the following format:  

                              Year	Value
	Vacant Housing Units	  -----   -----
	Occupied Housing Units	-----   -----

(d) (10 pts) Calculate the harmonic mean of the total housing units and print it out. You can find the information about harmonic mean here: https://ncalculators.com/statistics/harmonic-mean-calculator.htm

In [ ]:
import pandas as pd
# (a)
df = pd.read_csv('Assignment1_denton_housing.csv')
pivot_df = df.pivot(index='year', columns='title_field', values='value').reset_index()
pivot_df['Difference'] = pivot_df['Occupied Housing Units'].diff()
pivot_df = pivot_df.dropna(subset=['Difference'])
print("Difference in the number of Occupied Housing Units from year to year")
print(pivot_df['Difference'].sort_values(ascending=True))


# (b)
pivot_df = df.pivot(index='year', columns='title_field', values='value').reset_index()

pivot_df['Vacant Percentage'] = (pivot_df['Vacant Housing Units'] / pivot_df['Total Housing Units']) * 100
pivot_df['Occupied Percentage'] = (pivot_df['Occupied Housing Units'] / pivot_df['Total Housing Units']) * 100
print("Percentage of housing units which are vacant an occupied")
# Printing the results in the specified format
for index, row in pivot_df.iterrows():
    year = int(row['year'])
    vacant_percentage = f"{row['Vacant Percentage']:.0f}%"
    occupied_percentage = f"{row['Occupied Percentage']:.0f}%"
    print(f"{year} {vacant_percentage} {occupied_percentage}")


# (c)
pivot_df = df.pivot(index='year', columns='title_field', values='value').reset_index()
# finding maximum vecant and maximum occupied housing units.
max_vacant = pivot_df.loc[pivot_df['Vacant Housing Units'].idxmax()]
max_occupied = pivot_df.loc[pivot_df['Occupied Housing Units'].idxmax()]

print(f"Vacant Housing Units      {max_vacant['year']:>5}   {max_vacant['Vacant Housing Units']:>5}")
print(f"Occupied Housing Units    {max_occupied['year']:>5}   {max_occupied['Occupied Housing Units']:>5}")



# (d)
from scipy.stats import hmean
total_units = pivot_df['Total Housing Units']
# Calculatinh hmean by using built-in function
harmonic_mean = hmean(total_units[total_units > 0])
print(f"Harmonic Mean of Total Housing Units: {harmonic_mean:.2f}")


Difference in the number of Occupied Housing Units from year to year
1    -795.0
3    -303.0
4    1404.0
5    1565.0
6    2622.0
2    5091.0
Name: Difference, dtype: float64
Percentage of housing units which are vacant an occupied
2009 9% 91%
2010 8% 92%
2011 9% 91%
2012 7% 93%
2013 9% 91%
2014 6% 94%
Vacant Housing Units       2008    4483
Occupied Housing Units     2014   46295
Harmonic Mean of Total Housing Units: 44398.29


# Mandatory Question

Provide your thoughts on the assignment. What did you find challenging, and what aspects did you enjoy? Your opinion on the provided time to complete the assignment. Share any suggestions for improvement. Your feedback will be valuable for enhancing the learning experience.

In [ ]:
# Write your response below
I found the data science part a bit challenging. I think the provided time is suffucient for this assignment. I enjoyed working on this assignment.